tools/build_pytorch_libs.py

```diff
@@ -0,0 +1,283 @@
+        # package_data takes a map from 'python package' to 'globs of files to
+        # include'. By 'python package', it means a folder with an __init__.py file
+        # that's not excluded in the find_packages call earlier in setup.py. So to
+        # include our cpp_test into the site-packages folder in
+        # site-packages/caffe2/cpp_test, we have to copy the cpp_test folder into the
+        # root caffe2 folder and then tell setup.py to include them. Having another
+        # folder like site-packages/caffe2_cpp_test would also be possible by adding a
+        # caffe2_cpp_test folder to pytorch with an __init__.py in it.
+        mkdir_p(base_dir + '/caffe2/cpp_test/')
+        copy_tree(install_dir + '/test', base_dir + '/caffe2/cpp_test', update=True)
+
+
+def build_caffe2(version,
+                 cmake_python_library,
+                 build_python,
+                 rerun_cmake,
+                 build_dir):
+    build_test = not check_negative_env_flag('BUILD_TEST')
+    if rerun_cmake or not os.path.exists('build/CMakeCache.txt'):
+        run_cmake(version,
+                  cmake_python_library,
+                  build_python,
+                  build_test,
+                  build_dir)
+
+    if IS_WINDOWS:
+        if USE_NINJA:
+            # sccache will fail if all cores are used for compiling
+            j = max(1, multiprocessing.cpu_count() - 1)
+            check_call(['cmake', '--build', '.', '--target', 'install', '--config', build_type, '--', '-j', str(j)],
+                       cwd=build_dir)
+        else:
+            check_call(['msbuild', 'INSTALL.vcxproj', '/p:Configuration={}'.format(build_type)],
+                       cwd=build_dir)
+    else:
+        if USE_NINJA:
+            check_call(['ninja', 'install'], cwd=build_dir)
+        else:
+            check_call(['make', '-j', str(multiprocessing.cpu_count()), 'install'], cwd=build_dir)
+
+    # in cmake, .cu compilation involves generating certain intermediates
+    # such as .cu.o and .cu.depend, and these intermediates finally get compiled
+    # into the final .so.
+    # Ninja updates build.ninja's timestamp after all dependent files have been built,
+    # and re-kicks cmake on incremental builds if any of the dependent files
+    # have a timestamp newer than build.ninja's timestamp.
+    # There is a cmake bug with the Ninja backend, where the .cu.depend files
+    # are still compiling by the time the build.ninja timestamp is updated,
+    # so the .cu.depend file's newer timestamp is screwing with ninja's incremental
+    # build detector.
+    # This line works around that bug by manually updating the build.ninja timestamp
+    # after the entire build is finished.
+    if os.path.exists('build/build.ninja'):
+        os.utime('build/build.ninja', None)
+
+    if build_python:
+        for proto_file in glob('build/caffe2/proto/*.py'):
+            if proto_file != 'build/caffe2/proto/__init__.py':
```

Comment:
This one won't work on Windows. The output of glob will contain backslashes instead of slashes. It passes build just because `importing caffe2.python` isn't tested on Windows.